### PR TITLE
Fix UserCodeDeployment to work with inner classes

### PR DESCRIPTION
**Pull request by sancar**
The problem with only core side user code deployment.
Tests are added to the client as well.

Also cleaned up UserCodeDeploymentAbstractTest which
were used for the compatibility tests on enterprise repo.
Related cleanup pr https://github.com/hazelcast/hazelcast-enterprise/pull/3787

Compatibility tests are removed since they are not needed
anymore after switching to 4.0.

Following two tests are newly added to `UserCodeDeploymentBasicTest` which are related to this fix
testInnerClassFetchedFirst_thenMainClassFetchedFromRemote
testMainClassFetchedFirst_thenInnerlassFetchedFromRemote

Others are moved from `UserCodeDeploymentAbstractTest` when refactoring.

fixes https://github.com/hazelcast/hazelcast/issues/17044

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/internal/usercodedeployment/impl/filter/UserCodeDeploymentBasicTest.java [hunk 1]
```diff
@@ -19,28 +19,41 @@
...
+import static com.hazelcast.test.starter.HazelcastStarterUtils.assertInstanceOfByClassName;
 import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
 
 @RunWith(Parameterized.class)
 @Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
 @Category({QuickTest.class, ParallelJVMTest.class})
-public class UserCodeDeploymentBasicTest extends UserCodeDeploymentAbstractTest {
```
Comment (puzpuzpuz) [inline hazelcast/src/test/java/com/hazelcast/internal/usercodedeployment/impl/filter/UserCodeDeploymentBasicTest.java:42]: `UserCodeDeploymentAbstractTest` class is used by EE repo, so we should probably keep it.

Comment (puzpuzpuz) [reply hazelcast/src/test/java/com/hazelcast/internal/usercodedeployment/impl/filter/UserCodeDeploymentBasicTest.java:42]: Did see the note in the PR description. My bad. Resolving this one.

**Review by puzpuzpuz — APPROVED** [review]

File: hazelcast/src/test/java/usercodedeployment/DomainClassWithInnerClass.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    public static class InnerClass implements Serializable {
+        public final int value;
+
+        public InnerClass(int value) {
+            this.value = value;
+        }
+    }
+
+}
+
```
Comment (puzpuzpuz) [inline hazelcast/src/test/java/usercodedeployment/DomainClassWithInnerClass.java:38]: nit: redundant new line

**Review by kwart — APPROVED** [review]
LGTM.

**Pull request merged**